x-pack/test/functional/page_objects/security_page.ts

```diff
@@ -17,53 +18,114 @@ export function SecurityPageProvider({ getService, getPageObjects }) {
+
+  async function waitForLoginPage() {
+    await retry.waitForWithTimeout('login page', config.get('timeouts.waitFor') * 5, async () => {
+      const alert = await browser.getAlert();
+      if (alert && alert.accept) {
+        await alert.accept();
+      }
+      return await find.existsByDisplayedByCssSelector('.login-form');
+    });
+  }
+
+  async function waitForLoginForm() {
+    log.debug('Waiting for Login Form to appear.');
+    await retry.waitForWithTimeout('login form', config.get('timeouts.waitFor') * 5, async () => {
+      const alert = await browser.getAlert();
+      if (alert && alert.accept) {
+        await alert.accept();
+      }
+
+      return await testSubjects.exists('loginForm');
+    });
+  }
 
-      username = username || superUsername;
-      password = password || superPassword;
+  async function waitForLoginSelector() {
+    log.debug('Waiting for Login Selector to appear.');
+    await retry.waitForWithTimeout(
+      'login selector',
+      config.get('timeouts.waitFor') * 5,
+      async () => {
+        return await testSubjects.exists('loginSelector');
+      }
+    );
+  }
+
+  async function waitForLoginHelp(helpText: string) {
+    log.debug(`Waiting for Login Help to appear with text: ${helpText}.`);
+    await retry.waitForWithTimeout('login help', config.get('timeouts.waitFor') * 5, async () => {
+      return (await testSubjects.getVisibleText('loginHelp')) === helpText;
+    });
+  }
+
+  async function waitForLoginResult(expectedResult?: LoginExpectedResult) {
+    log.debug(`Waiting for login result, expected: ${expectedResult}.`);
+
+    // wait for either space selector, kibanaChrome or loginErrorMessage
+    if (expectedResult === 'spaceSelector') {
+      await retry.try(() => testSubjects.find('kibanaSpaceSelector'));
+      log.debug(
+        `Finished login process, landed on space selector. currentUrl = ${await browser.getCurrentUrl()}`
+      );
+      return;
+    }
 
-      const expectSpaceSelector = options.expectSpaceSelector || false;
-      const expectSuccess = options.expectSuccess;
-      const expectForbidden = options.expectForbidden || false;
+    if (expectedResult === 'error') {
       const rawDataTabLocator = 'a[id=rawdata-tab]';
+      if (await find.existsByCssSelector(rawDataTabLocator)) {
+        // Firefox has 3 tabs and requires navigation to see Raw output
+        await find.clickByCssSelector(rawDataTabLocator);
+      }
+      await retry.try(async () => {
+        if (await find.existsByCssSelector(rawDataTabLocator)) {
+          await find.clickByCssSelector(rawDataTabLocator);
+        }
+        await PageObjects.error.expectForbidden();
+      });
+      log.debug(
+        `Finished login process, found forbidden message. currentUrl = ${await browser.getCurrentUrl()}`
+      );
+      return;
+    }
 
+    if (expectedResult === 'chrome') {
+      await find.byCssSelector('[data-test-subj="kibanaChrome"] nav:not(.ng-hide) ', 20000);
```

Comment:
**question** The Kibana chrome isn't still using angular, is it? `ng-hide` probably won't ever exist now that it's migrated to KP. I know this was just a TS conversion, and we have this in other places too, but I wonder what the "correct" test should be. Maybe we just don't need that pseudo-selector anymore?

Reply:
That's a good point, let me figure that out while I'm here.

Reply:
It seems what we really need is `[data-test-subj="kibanaChrome"] .app-wrapper:not(.hidden-chrome)` - changed to this instead.